File: proyecto/go-side/internal/router/control/messages.go
// Package control implementa el plano de control: HELLO, LSA, flooding y tabla de ruteo.
package control

// Tipos de mensaje del protocolo (docs/protocol.md).
const (
	HELLO = "HELLO"
	LSA   = "LSA"
	DATA  = "DATA"
)

// Hello es el mensaje periodico que confirma que un vecino esta vivo.
type Hello struct {
	Type string `json:"type"`
	From string `json:"from"`
}

// BuildHello arma un HELLO desde el identificador propio (ip:puerto).
func BuildHello(fromID string) Hello { return Hello{Type: HELLO, From: fromID} }

// LSAMessage es el Link State Advertisement que se inunda a toda la red.
type LSAMessage struct {
	Type  string         `json:"type"`
	Origin string        `json:"origin"`
	Seq   int            `json:"seq"`
	Links map[string]int `json:"links"`
	From  string         `json:"from"`
}

// BuildLSA arma un LSA propio; origin y from inician iguales al emisor.
func BuildLSA(origin string, seq int, links map[string]int, fromID string) LSAMessage {
	return LSAMessage{Type: LSA, Origin: origin, Seq: seq, Links: links, From: fromID}
}

// DataEnvelope transporta el frame {from,to,msg} protegido con Hamming(7,4)
// completo (ver forwarding.Forward): un router intermedio decodifica TODO
// el frame pero solo lee "to"; unicamente el destino final usa "msg".
type DataEnvelope struct {
	Type string `json:"type"`
	Len  int    `json:"len"`
	Bits string `json:"bits"`
}

// Payload es el contenido logico protegido dentro de un DataEnvelope.
type Payload struct {
	From string `json:"from"`
	To   string `json:"to"`
	Msg  string `json:"msg"`
}

File: proyecto/python-side/router/control/messages.py
"""Construccion y parseo de los mensajes JSON del protocolo (docs/protocol.md)."""

from __future__ import annotations

import json

HELLO = "HELLO"
LSA = "LSA"
DATA = "DATA"


def build_hello(from_ip: str) -> dict:
    return {"type": HELLO, "from": from_ip}


def build_lsa(origin: str, seq: int, links: dict[str, int], from_ip: str) -> dict:
    return {"type": LSA, "origin": origin, "seq": seq, "links": links, "from": from_ip}


def loads(line: str) -> dict:
    """Parsea una linea recibida por el socket (delimitada por \\n)."""
    return json.loads(line)


def dumps(message: dict) -> str:
    """Serializa un mensaje agregando el delimitador de linea del protocolo."""
    return json.dumps(message) + "\n"
